proposals/generate_validating_admission_policies.md

```diff
@@ -0,0 +1,474 @@
+# Generate Validating Admission Policies
+
+- **Authors**: [Mariam Fahmy](https://github.com/MariamFahmy98)
+- **Created**: July 18th, 2023
+- **Abstract**: Generate validating admission policies from Kyverno `validate.cel` subrule.
+
+## Overview
+- [Introduction](#introduction)
+- [Motivation](#motivation)
+- [Proposal](#proposal)
+   + [Kyverno engine VS API server](#kyverno-engine-vs-api-server)
+   + [Limitations](#limitations)
+- [Implementation](#implementation)
+   + [ClusterPolicy vs Policy](#clusterpolicy-vs-policy)
+   + [Kyverno Policy Settings](#kyverno-policy-settings)
+   + [Match/Exclude Resources](#matchexclude-resources)
+   + [Validate Rules](#validate-rules)
+   + [Audit Annotations](#audit-annotations)
+   + [Parameter Resources](#parameter-resources)
+   + [Limitations](#limitations-1)
+- [Next Steps](#next-steps)
+
+## Introduction
+Kyverno policies support CEL expressions for resources validation. The goal is to generate Kubernetes validating admission policies and validating admission binding from the information provided by Kyverno policies.
+
+- In case a new Kyverno policy is created, a validating admission policy and its binding are generated. 
+- In case Kyverno policy is updated, the corresponding VAP and its binding needs to be updated accordingly. 
+- In case Kyverno policy is deleted, the corresponding VAP and its binding are deleted.
+
+Once validating admission policy and its binding are generated, the API server will handle the resource validation. 
+
+An example of Kyverno policy from which validating admission policy and its binding will be generated:
+```yaml=
+apiVersion: kyverno.io/v1
+kind: ClusterPolicy
+metadata:
+  name: check-deployment-replicas
+spec:
+  validationFailureAction: Enforce
+  background: false
+  rules:
+    - name: deployment-replicas
+      match:
+        any:
+        - resources:
+            kinds:
+              - Deployment
+      validate:
+        cel:
+          expressions:
+            - expression: "object.spec.replicas < 5"
+              message:  "Deployment spec.replicas must be less than 5”
+```
+
+## Motivation
+Currently, Kyverno engine supports CEL expressions for resource validation. However, validating admission policy is an alternative solution to validating admission webhooks. Therefore, it's better to make Kyverno generate validating admission policies so that resource validation will be handled by the API server instead.
+
+Users can either let Kyverno engine to validate resources using CEL expressions or let the API server to validate resources and this is done by allowing Kyverno to generate validating admission policies and their bindings from a Kyverno policy containing a CEL subrule.
+
+## Proposal
+- A new controller `validatingadmissionpolicy-controller` will be created. It will watch the following:
+  1. Kyverno policies
```

Comment:
all, or with a specific label?

Reply:
If the controller is enabled, then it will watch all Kyverno policies and check if it's `validate.CEL` subrule.